modules/reactor/src/meshgenerators/PinMeshGenerator.C

```diff
@@ -475,38 +430,105 @@ PinMeshGenerator::PinMeshGenerator(const InputParameters & parameters)
+
+  // Set metadata to describe pin attributes
+  declareMeshProperty(RGMB::pitch, _pitch);
+  declareMeshProperty(RGMB::is_homogenized, _homogenized);
+  declareMeshProperty(RGMB::is_single_pin, _is_assembly);
+  declareMeshProperty(RGMB::ring_radii, _ring_radii);
+  declareMeshProperty(RGMB::duct_halfpitches, _duct_halfpitch);
+  declareMeshProperty(RGMB::extruded, _extrude && _mesh_dimensions == 3);
+
+  unsigned int n_axial_levels =
+      (_mesh_dimensions == 3)
+          ? getReactorParam<std::vector<unsigned int>>(RGMB::axial_mesh_intervals).size()
+          : 1;
+  std::vector<std::vector<subdomain_id_type>> ring_region_ids(
+      n_axial_levels, std::vector<subdomain_id_type>(_ring_radii.size()));
+  std::vector<std::vector<subdomain_id_type>> duct_region_ids(
+      n_axial_levels, std::vector<subdomain_id_type>(_duct_halfpitch.size()));
+  std::vector<subdomain_id_type> background_region_ids(n_axial_levels);
+
+  for (const auto axial_idx : make_range(n_axial_levels))
   {
-    const auto node_id_background_meta =
-        getMeshProperty<dof_id_type>("node_id_background_meta", name() + "_2D");
-    setMeshProperty("node_id_background_meta", node_id_background_meta);
+    for (const auto ring_idx : index_range(_ring_radii))
+      ring_region_ids[axial_idx][ring_idx] = _region_ids[axial_idx][ring_idx];
+
+    background_region_ids[axial_idx] = _region_ids[axial_idx][_ring_radii.size()];
+
+    for (unsigned int duct_idx = _ring_radii.size() + 1;
+         duct_idx < _duct_halfpitch.size() + _ring_radii.size() + 1;
+         ++duct_idx)
+      duct_region_ids[axial_idx][duct_idx - _ring_radii.size() - 1] =
+          _region_ids[axial_idx][duct_idx];
   }
 
+  // Define mesh properties related to region ids
+  declareMeshProperty(RGMB::ring_region_ids, ring_region_ids);
+  declareMeshProperty(RGMB::background_region_id, background_region_ids);
+  declareMeshProperty(RGMB::duct_region_ids, duct_region_ids);
+
+  if (getParam<bool>("show_rgmb_metadata"))
+  {
+    std::string mg_struct = _is_assembly ? "assembly" : "pin";
+    printReactorMetadata(mg_struct, name());
+  }
+}
+
+std::unique_ptr<MeshBase>
+PinMeshGenerator::generate()
+{
+  // Must be called to free the ReactorMeshParams mesh
+  freeReactorMeshParams();
+
+  // Update metadata at this point since values for these metadata only get set by PCCMG
+  // at generate() stage
+  const auto max_radius_meta = getMeshProperty<Real>("max_radius_meta", name() + "_2D");
+  setMeshProperty("max_radius_meta", max_radius_meta);
+  const auto background_intervals_meta =
+      getMeshProperty<unsigned int>("background_intervals_meta", name() + "_2D");
+  setMeshProperty("background_intervals_meta", background_intervals_meta);
+  const auto node_id_background_meta =
+      getMeshProperty<dof_id_type>("node_id_background_meta", name() + "_2D");
+  setMeshProperty("node_id_background_meta", node_id_background_meta);
```

Comment:
can you use the copyMeshProperty here?

Reply:
Don't think so, we need to call setMeshProperty here since these calls occur in `generate` and not in the constructor

Reply:
We'd have to create a new function to accomplish this